src/csharp/Extensions/Microsoft.Spark.Extensions.DotNet.Interactive/AssemblyKernelExtension.cs

```diff
@@ -0,0 +1,155 @@
+            new PackageHelper(new PackageRestoreContextWrapper());
+
+        /// <summary>
+        /// Called by the Microsoft.DotNet.Interactive Assembly Extension Loader.
+        /// </summary>
+        /// <param name="kernel">The kernel calling this method.</param>
+        /// <returns><see cref="Task.CompletedTask"/> when extension is loaded.</returns>
+        public Task OnLoadAsync(IKernel kernel)
+        {
+            if (kernel is CompositeKernel kernelBase)
+            {
+                Environment.SetEnvironmentVariable(Constants.RunningREPLEnvVar, "true");
+
+                DirectoryInfo tempDir = CreateTempDirectory();
+                kernelBase.RegisterForDisposal(new DisposableDirectory(tempDir));
+
+                kernelBase.AddMiddleware(async (command, context, next) =>
+                {
+                    if ((context.HandlingKernel is CSharpKernel kernel) &&
+                        (command is SubmitCode) &&
+                        TryGetSparkSession(out SparkSession sparkSession) &&
+                        TryEmitAssembly(kernel, tempDir.FullName, out string assemblyPath))
+                    {
+                        sparkSession.SparkContext.AddFile(assemblyPath);
+
+                        foreach (string filePath in GetPackageFiles(tempDir.FullName))
+                        {
+                            sparkSession.SparkContext.AddFile(filePath);
+                        }
+                    }
+
+                    await next(command, context);
+                });
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private DirectoryInfo CreateTempDirectory()
+        {
+            string envTempDir = Environment.GetEnvironmentVariable(TempDirEnvVar);
+            string tempDirBasePath = string.IsNullOrEmpty(envTempDir) ?
+                Directory.GetCurrentDirectory() :
+                envTempDir;
+
+            if (!IsPathValid(tempDirBasePath))
+            {
+                throw new Exception($"[{GetType().Name}] Spaces in " +
+                    $"'{tempDirBasePath}' is unsupported. Set the {TempDirEnvVar} " +
+                    "environment variable to control the base path. Please see " +
+                    "https://issues.apache.org/jira/browse/SPARK-30126 and " +
+                    "https://github.com/apache/spark/pull/26773 for more details");
+            }
+
+            return Directory.CreateDirectory(
+                Path.Combine(tempDirBasePath, Path.GetRandomFileName()));
+        }
+
+        private bool TryEmitAssembly(CSharpKernel kernel, string dstPath, out string assemblyPath)
+        {
+            Compilation compilation = kernel.ScriptState.Script.GetCompilation();
+            string assemblyName =
+                AssemblyLoader.NormalizeAssemblyName(compilation.AssemblyName);
+            assemblyPath = Path.Combine(dstPath, $"{assemblyName}.dll");
+            if (!File.Exists(assemblyPath))
```

Comment:
I thought this was a critical error scenario no?

Reply:
I don't think it's expected.   Should we log an error or throw an exception ?

Reply:
let's throw an exception to fail fast